module/VuFind/src/VuFind/Controller/MyResearchController.php

```diff
@@ -1212,6 +1212,127 @@ public function checkedoutAction()
     }
 
     /**
+     * Send list of historic loans to view
+     *
+     * @return mixed
+     */
+    public function historicloansAction()
+    {
+        // Stop now if the user does not have valid catalog credentials available:
+        if (!is_array($patron = $this->catalogLogin())) {
+            return $patron;
+        }
+
+        // Connect to the ILS:
+        $catalog = $this->getILS();
+
+        // Check function config
+        $functionConfig = $catalog->checkFunction(
+            'getMyTransactionHistory', $patron
+        );
+        if (false === $functionConfig) {
+            $this->flashMessenger()->addErrorMessage('ils_action_unavailable');
+            return $this->createViewModel();
+        }
+
+        // Get page and page size:
+        $page = (int)$this->params()->fromQuery('page', 1);
+        $config = $this->getConfig();
+        $limit = isset($config->Catalog->historic_loan_page_size)
+            ? $config->Catalog->historic_loan_page_size : 50;
+        $ilsPaging = true;
+        if (isset($functionConfig['max_results'])) {
+            $limit = min([$functionConfig['max_results'], $limit]);
+        } elseif (isset($functionConfig['page_size'])) {
+            if (!in_array($limit, $functionConfig['page_size'])) {
+                $limit = isset($functionConfig['default_page_size'])
+                    ? $functionConfig['default_page_size']
+                    : $functionConfig['page_size'][0];
+            }
+        } else {
+            $ilsPaging = false;
+        }
+
+        // Get sort settings
+        $sort = false;
+        if (!empty($functionConfig['sort'])) {
+            $sort = $this->params()->fromQuery('sort');
+            if (!isset($functionConfig['sort'][$sort])) {
+                if (isset($functionConfig['default_sort'])) {
+                    $sort = $functionConfig['default_sort'];
+                } else {
+                    reset($functionConfig['sort']);
+                    $sort = key($functionConfig['sort']);
+                }
+            }
+        }
+
+        // Configure call params
+        $params = [
+            'sort' => $sort
+        ];
+        if ($ilsPaging) {
+            $params['page'] = $page;
+            $params['limit'] = $limit;
+        }
+
+        // Get checked out item details:
+        $result = $catalog->getMyTransactionHistory($patron, $params);
+
+        // Build paginator if needed:
+        if ($ilsPaging && $limit < $result['count']) {
```

Comment:
I'm guessing that some of this logic is copied and pasted from elsewhere. Might it make sense to create a support method for building paginators in order to reduce redundancy? Perhaps some other parts of this long method could also be broken out into support methods.

Reply:
There's a similar block in checkedoutAction(), but since it doesn't currently have support for ILS-provided paging, it's not a direct copy. I suppose it would be possible to create the pager in a support method but I'm not sure if that makes it cleaner at all. I can try it out if you wish. As far as I can see there are no other places where a similar pager is created.

Reply:
We can leave it alone for now. Ultimately I'd like to split MyResearchController into smaller parts. There's too much going on in there, and too many long functions... but adding one more won't make it significantly worse, and we can address the refactoring later.